Detect `AString.clone()` when `&str` is needed and emit appropriate suggestion

On type mismatch with a suggestion to borrow in order to dereference a `String` into a `&str`, we should notice unnecessary `.clone()` calls:

```
error[E0308]: mismatched types
 --> src/main.rs:7:38
  |
7 |     assert_eq!(make_lipographic('e', passage.clone()), passage);
  |                                      ^^^^^^^^^^^^^^^
  |                                      |
  |                                      expected &str, found struct `std::string::String`
  |                                      help: consider borrowing here: `&passage.clone()`
  |
  = note: expected type `&str`
             found type `std::string::String`
```

and suggest their removal:

```
error[E0308]: mismatched types
 --> src/main.rs:7:38
  |
7 |     assert_eq!(make_lipographic('e', passage.clone()), passage);
  |                                      ^^^^^^^^^^^^^^^
  |                                      |
  |                                      expected &str, found struct `std::string::String`
  |                                      help: consider borrowing and avoiding the clone here: `&passage`
  |
  = note: expected type `&str`
             found type `std::string::String`
```

_From https://thenewwazoo.github.io/clone.html_

--- a/src/librustc_typeck/check/demand.rs
+++ b/src/librustc_typeck/check/demand.rs
@@ -425,6 +425,9 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                                 }
                             }
                         }
+                        // If this expression had a clone call, when suggesting borrowing, we
+                        // want to suggest removing it
+                        let sugg_expr = sugg_expr.trim_end_matches(".clone()");
                         return Some(match mutability {
                             hir::Mutability::MutMutable => (
                                 sp,

--- a/src/librustc_typeck/check/demand.rs
+++ b/src/librustc_typeck/check/demand.rs
@@ -425,19 +425,29 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                                 }
                             }
                         }
-                        // If this expression had a clone call, when suggesting borrowing, we
-                        // want to suggest removing it
-                        let sugg_expr = sugg_expr.trim_end_matches(".clone()");
+
+                        let mut sugg = sugg_expr.as_str();
+                        if let hir::ExprKind::MethodCall(_segment, _sp, _args) = &expr.node {
+                            let clone_path = "std::clone::Clone::clone";
+                            if let Some(true) = self.tables.borrow()
+                                .type_dependent_def_id(expr.hir_id)
+                                .map(|did| self.tcx.def_path_str(did).as_str() == clone_path)
+                            {
+                                // If this expression had a clone call when suggesting borrowing
+                                // we want to suggest removing it because it'd now be unecessary.
+                                sugg = sugg_expr.trim_end_matches(".clone()");
+                            }
+                        }
                         return Some(match mutability {
                             hir::Mutability::MutMutable => (
                                 sp,
                                 "consider mutably borrowing here",
-                                format!("{}&mut {}", field_name, sugg_expr),
+                                format!("{}&mut {}", field_name, sugg),
                             ),
                             hir::Mutability::MutImmutable => (
                                 sp,
                                 "consider borrowing here",
-                                format!("{}&{}", field_name, sugg_expr),
+                                format!("{}&{}", field_name, sugg),
                             ),
                         });
                     }

--- a/src/librustc_typeck/check/demand.rs
+++ b/src/librustc_typeck/check/demand.rs
@@ -380,12 +380,16 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                 };
                 if self.can_coerce(ref_ty, expected) {
                     let mut sugg_sp = sp;
-                    if let hir::ExprKind::MethodCall(_segment, _sp, args) = &expr.node {
-                        let clone_path = "std::clone::Clone::clone";
-                        if let ([arg], Some(true)) = (&args[..], self.tables.borrow()
-                            .type_dependent_def_id(expr.hir_id)
-                            .map(|did| self.tcx.def_path_str(did).as_str() == clone_path))
-                        {
+                    if let hir::ExprKind::MethodCall(segment, _sp, args) = &expr.node {
+                        let clone_trait = self.tcx.lang_items().clone_trait().unwrap();
+                        if let ([arg], Some(true), "clone") = (
+                            &args[..],
+                            self.tables.borrow().type_dependent_def_id(expr.hir_id).map(|did| {
+                                let ai = self.tcx.associated_item(did);
+                                ai.container == ty::TraitContainer(clone_trait)
+                            }),
+                            &segment.ident.as_str()[..],
+                        ) {
                             // If this expression had a clone call when suggesting borrowing
                             // we want to suggest removing it because it'd now be unecessary.
                             sugg_sp = arg.span;

--- a/src/librustc_typeck/check/demand.rs
+++ b/src/librustc_typeck/check/demand.rs
@@ -379,7 +379,19 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                     }
                 };
                 if self.can_coerce(ref_ty, expected) {
-                    if let Ok(src) = cm.span_to_snippet(sp) {
+                    let mut sugg_sp = sp;
+                    if let hir::ExprKind::MethodCall(_segment, _sp, args) = &expr.node {
+                        let clone_path = "std::clone::Clone::clone";
+                        if let ([arg], Some(true)) = (&args[..], self.tables.borrow()
+                            .type_dependent_def_id(expr.hir_id)
+                            .map(|did| self.tcx.def_path_str(did).as_str() == clone_path))
+                        {
+                            // If this expression had a clone call when suggesting borrowing
+                            // we want to suggest removing it because it'd now be unecessary.
+                            sugg_sp = arg.span;
+                        }
+                    }
+                    if let Ok(src) = cm.span_to_snippet(sugg_sp) {
                         let needs_parens = match expr.node {
                             // parenthesize if needed (Issue #46756)
                             hir::ExprKind::Cast(_, _) |
@@ -426,28 +438,16 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                             }
                         }
 
-                        let mut sugg = sugg_expr.as_str();
-                        if let hir::ExprKind::MethodCall(_segment, _sp, _args) = &expr.node {
-                            let clone_path = "std::clone::Clone::clone";
-                            if let Some(true) = self.tables.borrow()
-                                .type_dependent_def_id(expr.hir_id)
-                                .map(|did| self.tcx.def_path_str(did).as_str() == clone_path)
-                            {
-                                // If this expression had a clone call when suggesting borrowing
-                                // we want to suggest removing it because it'd now be unecessary.
-                                sugg = sugg_expr.trim_end_matches(".clone()");
-                            }
-                        }
                         return Some(match mutability {
                             hir::Mutability::MutMutable => (
                                 sp,
                                 "consider mutably borrowing here",
-                                format!("{}&mut {}", field_name, sugg),
+                                format!("{}&mut {}", field_name, sugg_expr),
                             ),
                             hir::Mutability::MutImmutable => (
                                 sp,
                                 "consider borrowing here",
-                                format!("{}&{}", field_name, sugg),
+                                format!("{}&{}", field_name, sugg_expr),
                             ),
                         });
                     }

--- a/src/test/ui/issues/issue-61106.rs
+++ b/src/test/ui/issues/issue-61106.rs
@@ -1,0 +1,6 @@
+fn main() {
+    let x = String::new();
+    foo(x.clone()); //~ ERROR mismatched types
+}
+
+fn foo(_: &str) {}

--- a/src/test/ui/issues/issue-61106.stderr
+++ b/src/test/ui/issues/issue-61106.stderr
@@ -1,0 +1,15 @@
+error[E0308]: mismatched types
+  --> $DIR/issue-61106.rs:3:9
+   |
+LL |     foo(x.clone());
+   |         ^^^^^^^^^
+   |         |
+   |         expected &str, found struct `std::string::String`
+   |         help: consider borrowing here: `&x`
+   |
+   = note: expected type `&str`
+              found type `std::string::String`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0308`.

